yarn-project/ethereum/src/l1_tx_utils.ts

```diff
@@ -0,0 +1,309 @@
+              if (err instanceof Error && err.message.includes('reverted')) {
+                throw err;
+              }
+            }
+          }
+        }
+
+        const tx = await this.publicClient.getTransaction({ hash: currentTxHash });
+        const timePassed = Date.now() - lastSeen;
+
+        if (tx && timePassed < gasConfig.stallTimeMs!) {
+          this.logger?.debug(`L1 Transaction ${currentTxHash} pending. Time passed: ${timePassed}ms`);
+          lastSeen = Date.now();
+          await sleep(gasConfig.checkIntervalMs!);
+          continue;
+        }
+
+        if (timePassed > gasConfig.stallTimeMs! && attempts < gasConfig.maxAttempts!) {
+          attempts++;
+          const newGasPrice = await this.getGasPrice(gasConfig, attempts);
+
+          this.logger?.debug(
+            `L1 Transaction ${currentTxHash} appears stuck. Attempting speed-up ${attempts}/${gasConfig.maxAttempts} ` +
+              `with new priority fee ${formatGwei(newGasPrice.maxPriorityFeePerGas)} gwei`,
+          );
+
+          currentTxHash = await this.walletClient.sendTransaction({
+            ...request,
+            nonce,
+            gas: params.gasLimit,
+            maxFeePerGas: newGasPrice.maxFeePerGas,
+            maxPriorityFeePerGas: newGasPrice.maxPriorityFeePerGas,
+          });
+
+          txHashes.add(currentTxHash);
+          lastSeen = Date.now();
+        }
+        await sleep(gasConfig.checkIntervalMs!);
+      } catch (err: any) {
+        this.logger?.warn(`Error monitoring tx ${currentTxHash}:`, err);
+        if (err.message?.includes('reverted')) {
+          throw err;
+        }
+        await sleep(gasConfig.checkIntervalMs!);
+      }
+    }
+  }
+
+  /**
+   * Sends a transaction and monitors it until completion
+   * @param request - The transaction request (to, data, value)
+   * @param gasConfig - Optional gas configuration
+   * @returns The receipt of the successful transaction
+   */
+  public async sendAndMonitorTransaction(
+    request: L1TxRequest,
+    gasConfig?: Partial<L1TxUtilsConfig>,
+  ): Promise<TransactionReceipt> {
+    const { txHash, gasLimit } = await this.sendTransaction(request, gasConfig);
+    return this.monitorTransaction(request, txHash, { gasLimit }, gasConfig);
+  }
+
+  /**
+   * Gets the current gas price with bounds checking
+   */
+  private async getGasPrice(_gasConfig?: L1TxUtilsConfig, attempt: number = 0): Promise<GasPrice> {
+    const gasConfig = { ...this.config, ..._gasConfig };
+    const block = await this.publicClient.getBlock({ blockTag: 'latest' });
+    const baseFee = block.baseFeePerGas ?? 0n;
+
+    // Get initial priority fee from the network
+    let priorityFee = await this.publicClient.estimateMaxPriorityFeePerGas();
+    if (attempt > 0) {
+      // Bump priority fee by configured percentage for each attempt
+      priorityFee = (priorityFee * (100n + (gasConfig.priorityFeeBumpPercentage ?? 20n) * BigInt(attempt))) / 100n;
+    }
```

Comment:
Heads up that this may return a priorityFee that's _lower_ than the one used in the previous attempt, if there was a sudden drop in the current max priority fee estimated by the public client. 

If that happens, the node will reject the tx replacement as "underpriced" when we try to send it. Not just that, but the node also requires that there is a minimum bump.

Reply:
Looking at geth's code, there is a minimum bump enforced for replacements for both the priority fee and the max fee per gas. So we'll need to ensure both are bumped wrt the previous tx we had sent. This means we cannot set the maxFee to the config maximum as I had suggested originally (apologies) since we would not be able to bump it at all if that were the case.

https://github.com/ethereum/go-ethereum/blob/e3d61e6db028c412f74bc4d4c7e117a9e29d0de0/core/txpool/legacypool/list.go#L298-L345

Reply:
And to make things even more annoying, it seems the replacement threshold is different for blob and regular txs. 

https://github.com/ethereum/go-ethereum/blob/e3d61e6db028c412f74bc4d4c7e117a9e29d0de0/core/txpool/blobpool/blobpool.go#L163-L168

Reply:
So, to consolidate: we need to at least bump both the maxFeePerGas and the maxPrioFeePerGas by a percentage wrt the previous tx we had sent (ie the one we want to replace), but also we want to take into account the current base fee and priority fee we see on the chain.